### ISSUE TITLE
Failure when using w2w with global LCZ map for WRF

### ISSUE THREAD

**Issue by LluisFB** [matthiasdemuzere/w2w#98]
### Describe your issue

I had been trying to use the WUDAPT data into my WRF simulations. When trying to use the specific LCZ maps (.tiff) for a giving city, they are too small for the WRF domains and w2w does not work. Example for Buenos Aires
```
(...)
Check if LCZ domain extends WRF domain in all directions?
        ERROR: LCZ domain should be larger than WRF domain in all directions.
        LCZ bounds  (xmin, ymin, xmax, ymax): (-58.86121067664753, -34.99477020816008, -58.11111741440773, -34.09106503233584)
        WRF bounds  (xmin, ymin, xmax, ymax): (-59.325897216796875, -35.270755767822266, -57.71136474609375, -33.981346130371094)
```
Then there is this new [global LCZ](https://zenodo.org/record/6364594/files/lcz_filter_v1.tif?download=1) data-set: [lcz-maps](https://www.wudapt.org/lcz-maps/) and I tried to use it.

> Demuzere, M., Kittner, J., Martilli, A., Mills, G., Moede, C., Stewart, I. D., van Vliet, J., and Bechtel, B. (2022). _A global map of Local Climate Zones to support earth system modelling and urban scale environmental science_, Earth Syst. Sci. Data **14(8)** 3835-3873. DOI: [10.5194/essd-14-3835-2022](https://doi.org/10.5194/essd-14-3835-2022).

But when using with w2w I got memory problems (from `top` I got about 20 GB memory use from w2w) and the program stopped. Then I moved to a machine with larger memory, but neither worked. It just stopped without error message

* Standard output
```
$ /w2w ./ lcz_filter_v1.tif geo_em.d01.nc
(...)
Looping through urban grid pixels: 100%|██████████| 2314/2314 [00:42<00:00, 54.31it/s]
```
* Generated Files:
```
 2.9G Oct 17 15:18 map_clean.tif
 8.5M Oct 17 15:19 geo_em.d01_NoUrban.nc
```
I thought to cut the LCZ global map to a smaller area closer to my area of interest, but neither worked (**NOTE:** I am not an expert on GIS systems, so, I might have done something wrong, since reduced file is larger than the original one):
```
(...)
$ gdal_translate -projwin -15.0 63.0 11.0 39.0 -of GTiff lcz_filter_v1.tif lcz_filter_v1_EuropeParis.tif
```
The last option I tried, consisted in to reduce manually my WRF domain by cutting it to a smaller lon/lat box. If w2w does not require the exact projection of the map, this might work, since it will reduce the size of the domain. Then I could bring back the right values at the right WRF domain grids over the entire domain. It did not work neither, it crashed without any error message.

I do not know what else to do. I sent by email the `geo_em.d01.nc` and the `lcz_filter_v1_EuropeParis.tif` files

Any help will be very welcome.

**Wish-list**: Now that there is a global WUDAPT data-set, for WRF users might help if the data can be provided in the format used ([binary raster format](https://www2.mmm.ucar.edu/wrf/users/docs/user_guide_v4/v4.4/users_guide_chap3.html#_Writing_Static_Data)) to provide the ancillary data for WRF Prepocessing system. It uses an specific file format and it uses multiple files each one with a portion of the globe. However, as far as I know it only works with this following [projections](https://www2.mmm.ucar.edu/wrf/users/docs/user_guide_v4/v4.4/users_guide_chap3.html#_Description_of_index): lambert, polar, mercator, regular_ll, albers_nad83, or polar_wgs84 . As WRF user and non GIS expert I can offer my help if needed

Thanks, 

Lluís
lluis.fita [a] cima.fcen.uba.ar

### w2w --version

0.4.2

### nc-config --all

```console
This netCDF 4.4.1.1 has been built with the following features: 

  --cc        -> /usr/bin/cc
  --cflags    -> -I/usr/include -I/usr/include/hdf5/serial
  --libs      -> -L/usr/lib/x86_64-linux-gnu -L/usr/lib/x86_64-linux-gnu/hdf5/serial -lnetcdf -lhdf5_hl -lhdf5 -lpthread -lsz -lz -ldl -lm -lcurl

  --has-c++   -> no
  --cxx       -> 

  --has-c++4  -> no
  --cxx4      -> 

  --has-fortran-> yes
  --fc        -> gfortran
  --fflags    -> -I/usr/include
  --flibs     -> -L/usr/lib -lnetcdff -fPIE -pie -Wl,-z,relro -Wl,-z,now -lnetcdf -lnetcdf
  --has-f90   -> no
  --has-f03   -> yes

  --has-dap   -> yes
  --has-nc2   -> yes
  --has-nc4   -> yes
  --has-hdf5  -> yes
  --has-hdf4  -> no
  --has-logging-> yes
  --has-pnetcdf-> no
  --has-szlib -> no

  --prefix    -> /usr
  --includedir-> /usr/include
  --libdir    -> /usr/lib/x86_64-linux-gnu
  --version   -> netCDF 4.4.1.1
```

### Installed Packages

```console
I installed directly following instructions from w2w wiki:
$ pip install git+https://github.com/matthiasdemuzere/w2w
```

**Comment by matthiasdemuzere** [comment]
> **Wish-list**: Now that there is a global WUDAPT data-set, for WRF users might help if the data can be provided in the format used ([binary raster format](https://www2.mmm.ucar.edu/wrf/users/docs/user_guide_v4/v4.4/users_guide_chap3.html#_Writing_Static_Data)) to provide the ancillary data for WRF Prepocessing system. It uses an specific file format and it uses multiple files each one with a portion of the globe. However, as far as I know it only works with this following [projections](https://www2.mmm.ucar.edu/wrf/users/docs/user_guide_v4/v4.4/users_guide_chap3.html#_Description_of_index): lambert, polar, mercator, regular_ll, albers_nad83, or polar_wgs84 . As WRF user and non GIS expert I can offer my help if needed

Let me reply first to this part: we are currently working on having the global LCZ map as part of WPS. When ready, W2W will be obsolete, and a user will be able to ingest LCZ information directly via a flag in the WPS namelist! But this may still take some time, as such we hope W2W is able to fill this gap.

**Comment by matthiasdemuzere** [comment]
@LluisFB: Could you also share your geo_em files (all nests)? 

Related question: typically one would only introduce the LCZ information in the smallest nest, on the highest resolution. Yet as far as I can see, you want to introduce LCZ info in the parent nest (d01)? If this intended?

**Comment by LluisFB** [comment]
Thank you for the WPS effort!

We have a single domain d01, directly at 3 km. There is not need of nesting. Also, it would be preferable to have a domain large enough to keep the borders about 150 km away from the area of interest. We tried a smaller one at the beginning and we found a lot of 'noise' inside the domain.

Thanks

**Comment by matthiasdemuzere** [comment]
> I do not know what else to do. I sent by email the `geo_em.d01.nc` and the `lcz_filter_v1_EuropeParis.tif` files

Ha, this email I never received? Could you try again? Or link an cloud archive here that contains these files (in case they are too big)?

**Comment by matthiasdemuzere** [comment]
Okay, I got a bit further on a different machine (AND the latest version of w2w - 0.4.2 - I was too fast in the test above)

```
--> Set data, arguments and files 
> Using layer 0 of the LCZ GeoTIFF. Can be changed with -l (--lcz-band)
--> Check LCZ integrity, in terms of class labels, projection and extent
> LCZ labels as expected (1 to 17)
> LCZ provided as WGS84 (EPSG:4326)
> LCZ domain is covering WRF domain
--> Replace WRF MODIS urban LC with surrounding natural LC
Looping through urban grid pixels: 100%|██████████████████████████████████| 2314/2314 [00:21<00:00, 107.28it/s]
--> Create LCZ-based geo_em file
> Processing LP_URB2D ...
> Processing MH_URB2D ...
> Processing STDH_URB2D ...
> Processing HGT_URB2D ...
> Processing LB_URB2D ...
> Processing LF_URB2D ...
> Processing HI_URB2D ...
--> Create LCZ-based urban extent geo_em file (excluding other LCZ-based info)
--> Expanding land categories of parent domain(s) to 41
Traceback (most recent call last):
  File "/home/demuzmp4/data/w2w/Paris_EU_canicule/venv/bin/w2w", line 8, in <module>
    sys.exit(main())
  File "/home/demuzmp4/data/w2w/Paris_EU_canicule/venv/lib/python3.9/site-packages/w2w/w2w.py", line 195, in main
    expand_land_cat_parents(
  File "/home/demuzmp4/data/w2w/Paris_EU_canicule/venv/lib/python3.9/site-packages/w2w/w2w.py", line 1383, in expand_land_cat_parents
    domain_nr = int(info.dst_file[-5:-3])
ValueError: invalid literal for int() with base 10: '.1'
```

But there is an error when trying to expand the number of land categories in the parent domains (which you do not have).
I'll check this ...

But, the good news is that the required geo_em files are created: [geo_em.d01_v4.4.1_.zip](https://github.com/matthiasdemuzere/w2w/files/9821651/geo_em.d01_v4.4.1_.zip)

Please carefully check if these make sense ...

**Comment by LluisFB** [comment]
I check the files and they look fine.

So I guess the problem was two fold: version of w2w and size of the tiff file being used?

To get the latest version is enough to `pip install w2w`?

![ParisCanicule_lanuse3k_modif-IGBP-MODIS](https://user-images.githubusercontent.com/96058724/196784990-00bbbc45-ca81-49f1-a9a9-3583e46a87cf.png)

Many thanks!

Lluís

**Comment by LluisFB** [comment]
Could you please provide some more details on all your steps? I guess I will not be the unique one (or maybe yes) to face this kind of problems, or/and in order to be self-sufficient in other domains:
 a) How do you cut the global LCZ map?
 b) Why there was a need to update the tool w2w?

Thank you very much

**Comment by matthiasdemuzere** [comment]
Hi @LluisFB,

Some more feedback.

**1. LCZ map**
> gdal_translate -projwin -15.0 63.0 11.0 39.0 -of GTiff lcz_filter_v1.tif lcz_filter_v1_EuropeParis.tif

This works of course, yet your region of interest could be much smaller, and still extend the WRF domain, e.g.  -4 53 7.5 45.5.
That already reduces the cropped LCZ file.

In any case, differences between ways of cropping and resulting file sizes are interesting:
* Clipping from local asset, using gdal_translate, your domain: 773.4 MB
* Clipping from local asset, using gdal_translate, same domain: 106.9 MB
* Exporting domain  -4 53 7.5 45.5 from Google Earth Engine: LCZ map of 16MB


An then W2W cleans this tif, and stores a temporary _clean.tif version, used in the routine. This inflates the tif e.g. from 773.4MB to 3.1GB.

So I guess this is why you had the memory issue?

Notes:
- The LCZ map is now formally available as an asset in Google Earth Engine, see [here](https://developers.google.com/earth-engine/datasets/catalog/RUB_RUBCLIM_LCZ_global_lcz_map_v1).
- @theendlessriver13: we might have to check how to write a more compressed temporary _clean.tif, [here](https://github.com/matthiasdemuzere/w2w/blob/a90c686fb4bb8aa24e3b06c384b0761a0409e8f5/w2w/w2w.py#L399)?

**2. Bug when targeting parent domains?**
> --> Expanding land categories of parent domain(s) to 41
Traceback (most recent call last):
  File "/home/demuzmp4/data/w2w/Paris_EU_canicule/venv/bin/w2w", line 8, in <module>
    sys.exit(main())
  File "/home/demuzmp4/data/w2w/Paris_EU_canicule/venv/lib/python3.9/site-packages/w2w/w2w.py", line 195, in main
    expand_land_cat_parents(
  File "/home/demuzmp4/data/w2w/Paris_EU_canicule/venv/lib/python3.9/site-packages/w2w/w2w.py", line 1383, in expand_land_cat_parents
    domain_nr = int(info.dst_file[-5:-3])
ValueError: invalid literal for int() with base 10: '.1'

This error is not really a bug in the code. It is caused because your filename is `geo_em.d01_v4.4.1.nc`, while the W2W routine expects `geo_em.d01.nc`. [This section](https://github.com/matthiasdemuzere/w2w/blob/a90c686fb4bb8aa24e3b06c384b0761a0409e8f5/w2w/w2w.py#L1382) looks for the number before .nc, and then loops back over the previous numbers if they exist. 

**3. Executing W2W**

W2W - 0.4.2 (latest), your large LCZ crop, renaming `geo_em.d01_v4.4.1.nc` to `geo_em.d01.nc` (on server with more memory).


```
w2w . lcz_filter_v1_EuropeParis_large.tif geo_em.d01.nc
--> Set data, arguments and files 
> Using layer 0 of the LCZ GeoTIFF. Can be changed with -l (--lcz-band)
--> Check LCZ integrity, in terms of class labels, projection and extent
> LCZ labels as expected (1 to 17)
> LCZ provided as WGS84 (EPSG:4326)
> LCZ domain is covering WRF domain
--> Replace WRF MODIS urban LC with surrounding natural LC
Looping through urban grid pixels: 100%|██████████████████████████████████████| 2314/2314 [00:21<00:00, 109.71it/s]
--> Create LCZ-based geo_em file
> Processing LP_URB2D ...
> Processing MH_URB2D ...
> Processing STDH_URB2D ...
> Processing HGT_URB2D ...
> Processing LB_URB2D ...
> Processing LF_URB2D ...
> Processing HI_URB2D ...
--> Create LCZ-based urban extent geo_em file (excluding other LCZ-based info)
--> Expanding land categories of parent domain(s) to 41

--> Start sanity check and clean-up ...
> Check 1: Urban class removed from geo_em.d01_NoUrban.nc? OK 
> Check 2: LCZ Urban extent present in geo_em.d01_LCZ_extent.nc? OK 
> Check 3: Urban LCZ classes exists in geo_em.d01_LCZ_params.nc? OK: LCZ Classes ([1, 2, 3, 4, 5, 6, 8, 9, 10]) present 
> Check 4: FRC_URB2D present in geo_em.d01_LCZ_params.nc? OK: FRC_URB2D values range between 0.00 and 0.90 
> Check 5: URB_PARAMS matrix present in file geo_em.d01_LCZ_params.nc? OK 
> Check 6: Do URB_PARAM variable values follow expected range in geo_em.d01_LCZ_params.nc?
   + OK for LP_URB2D 
   + OK for MH_URB2D 
   + OK for STDH_URB2D 
   + OK for HGT_URB2D 
   + OK for LB_URB2D 
   + OK for LF_URB2D 
> Check 7: Does HI_URB2D sum to 100% for urban pixels in geo_em.d01_LCZ_params.nc? OK 
> Check 8: Do FRC_URB and LCZs (from LU_INDEX) cover same extent in geo_em.d01_LCZ_params.nc? OK 
> Check 9: Extent and # urban pixels same for *_extent.nc and *_params.nc output file? OK, urban extent the same (3791) 
> Cleaning up ... all done!


 ----------- !! NOTE !! --------- 
 Set nbui_max to 10 during compilation, in order to optimize memory storage.

```


In summary:
* I can not really replicate your issue
* There should not be a difference between version 0.4.1 and 0.4.2, since there are no relevant code changes between the two, see: [https://github.com/matthiasdemuzere/w2w/releases](https://github.com/matthiasdemuzere/w2w/releases) 
* It seems that the `_hi_resampler` [part of the code](https://github.com/matthiasdemuzere/w2w/blob/a90c686fb4bb8aa24e3b06c384b0761a0409e8f5/w2w/w2w.py#L951) consumes the most memory? We might want to look at that @theendlessriver13 in case people start using W2W for larger domains!
* Latest version of w2w can be installed with pip. Use `pip install -U w2w` to be sure.
